sql/schemachanger: implement `DROP COLUMN`

ajwerner opened the pull request:
See individual commits. 

Fixes #84072

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/84563)
<!-- Reviewable:end -->

ajwerner commented:
The last commit should be broken up into a few. I'll do that later after we see how CI is doing.

ajwerner commented:
Sent out the changefeed changes in https://github.com/cockroachdb/cockroach/pull/84571 and the testing fix in https://github.com/cockroachdb/cockroach/pull/84575.

postamar reviewed (commented):
I did a review pass on just the last commit. Nice work. This is certainly coming along! Very exciting.

> Reviewed 74 of 74 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7HCFBe4Bk4NdTChyun:bvx2xti)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @postamar)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7H5PDZ4ZG75acODlXs:-N7H5PDZ4ZG75acODlXt:b-pmxm7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f59c14b854b9fcc46fabc1d369014b1cac4e464/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L68)):*
> ```Go
> ) {
> 	var rowLevelTTL *scpb.RowLevelTTL
> 	// TODO(ajwerner): Does this need to look at status or target status?
> ```

I would look at status if only because target status may be changed by the builder. I have no idea how consistent we are about this. Looking ahead, we need a better way to express queries on current state.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7H8cwj3gGEG3oPMp3Q:-N7H8cwj3gGEG3oPMp3R:bjrhu9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f59c14b854b9fcc46fabc1d369014b1cac4e464/pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go#L68)):*
> ```Go
> 					return &scop.MakeMergedIndexWriteOnly{
> 						TableID: this.TableID,
> 						IndexID: this.IndexID,
> ```

I assume this extra op-edge was necessary for the rollbacks to work? What of the writes occurring between MERGED and WRITE_ONLY, won't they be lost?
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 830 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7HB9sk1DjVv6tyHXS0:-N7HB9sk1DjVv6tyHXS1:b-oi82rt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f59c14b854b9fcc46fabc1d369014b1cac4e464/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L830)):*
> ```Go
> 			}
> 		},
> 	)
> ```

I don't understand this rule. Shouldn't this be: the column can only reach WRITE_ONLY if the indexes can all be written to, in other words when the temp indexes are all WRITE_ONLY and the other indexes are also WRITE_ONLY?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
I've pushed again and addressed the concrete feedback. After the changefeed changes make their way through bors and CI is happy (which I think it should be), I'll take a stab at breaking this out into a few more commits. I do think it's quite close. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7HZSSV4LU5OnCO6h86:bgrzx9e)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7H5PDZ4ZG75acODlXs:-N7HZK17AB_vBqNqy3iS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f59c14b854b9fcc46fabc1d369014b1cac4e464/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L68)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I would look at status if only because target status may be changed by the builder. I have no idea how consistent we are about this. Looking ahead, we need a better way to express queries on current state.
</blockquote></details>

Done.
___
*[`pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go` line 68 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7H8cwj3gGEG3oPMp3Q:-N7HVkkm271FALGKIRtV:b-3tm0eb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f59c14b854b9fcc46fabc1d369014b1cac4e464/pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go#L68)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I assume this extra op-edge was necessary for the rollbacks to work? What of the writes occurring between MERGED and WRITE_ONLY, won't they be lost?
</blockquote></details>

This is necessary for very subtle reasons I remembered about only recently. We've had bugs we've been carrying around. I need to split this off into more commits to give some more motivation. This comment is revelatory: https://github.com/cockroachdb/cockroach/blob/116c5aae8c319a3cdb23e19adadf62b765e66f12/pkg/sql/backfill.go#L1962-L2039

The issue is that in `Merging` we do not enforce uniqueness constraints. We need to be enforcing uniqueness constraints before we go on to validating the uniqueness constraint. This enforcement happens through the use of a CPut, which we do not use in `Merging` here: https://github.com/cockroachdb/cockroach/blob/116c5aae8c319a3cdb23e19adadf62b765e66f12/pkg/sql/catalog/table_elements.go#L204-L221
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 830 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7HB9sk1DjVv6tyHXS0:-N7HVV6mAJVuXUSl0zip:b23v2fb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9f59c14b854b9fcc46fabc1d369014b1cac4e464/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L830)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

I don't understand this rule. Shouldn't this be: the column can only reach WRITE_ONLY if the indexes can all be written to, in other words when the temp indexes are all WRITE_ONLY and the other indexes are also WRITE_ONLY?
</blockquote></details>

I think the temp indexes condition would be sufficient. This one is also sufficient. The deal is that we need to make sure that we are writing values of the column somewhere before we start doing the backfill. I have rewritten the rule to be about the temporary index and added commentary. 


<!-- Sent from Reviewable.io -->

ajwerner committed:
sql/row: pretty print CPut expectation for kv trace

Release note: None

ajwerner committed:
sql/schemachanger: properly transition indexes to WRITE_ONLY, add rules

Release note: None

ajwerner committed:
sql,sqlerrors: extract some errors for reuse

Release note: None

ajwerner committed:
sql/rowenc: compute primary index encoding columns correctly

This commit ensures that the correct primary index entry is encoded when
building an entry for a primary index encoded index which does not include
every column in the table. This is the case when dropping a column using
the declarative schema changer.

Release note: None

ajwerner committed:
sql/logictest,execbuilder: update logic tests to avoid index IDs

These hard-coded index IDs embed assumptions which will be violated
by the new DROP COLUMN protocol.

Release note: None

ajwerner committed:
sql/schemachanger: implement `ALTER TABLE ... DROP COLUMN`

This is the initial implementation of `ALTER TABLE ... DROP COLUMN` in the
declarative schema changer. It explicitly does not have the case of adding
and dropping columns from the same table in the same transaction. It also
omits:

1) Dropping columns related to multi-region tables
2) Dropping columns with row-level TTLs
3) Dropping columns involved in UNIQUE WITHOUT INDEX constraints
4) Dropping columns involved in CHECK constraints
5) Dropping columns involved in FOREIGN KEY constraints

Fixes #84072.

Release note: None

ajwerner commented:
This is RFAL. I broke this apart into individual commits. I think that was a good exercise. Only really the fourth commit (
[sql/rowenc: compute primary index encoding columns correctly](https://github.com/cockroachdb/cockroach/pull/84563/commits/864b7d9c28c68286403926156965a641cfb90901)) is of real interest to outside teams. I was disappointed to discover the lack of testing in that area, but haven't done anything to make it better.

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @livlobo, and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1115 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7Ilp_57YvzQAtSDcTx:-N7Ilp_57YvzQAtSDcTy:b-moiyex) ([raw file](https://github.com/cockroachdb/cockroach/blob/864b7d9c28c68286403926156965a641cfb90901/pkg/sql/rowenc/index_encoding.go#L1115)):*
> ```Go
> 
> // getStoredColumnsForPrimaryIndex computes the set of columns stored in this
> // primary index's value for encoding which are not composite columns. The
> ```

So what do we do with composite columns then?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo, @mgartner, and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1115 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7Ilp_57YvzQAtSDcTx:-N7InK05-vM9B3uBLhUk:b-ar66sn) ([raw file](https://github.com/cockroachdb/cockroach/blob/864b7d9c28c68286403926156965a641cfb90901/pkg/sql/rowenc/index_encoding.go#L1115)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

So what do we do with composite columns then?
</blockquote></details>

Have a look at where the output of this thing is used. The composite logic remains intact. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo, @mgartner, and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1115 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7Ilp_57YvzQAtSDcTx:-N7InWF0B2ijpAqBF8fL:b7bf6qp) ([raw file](https://github.com/cockroachdb/cockroach/blob/864b7d9c28c68286403926156965a641cfb90901/pkg/sql/rowenc/index_encoding.go#L1115)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Have a look at where the output of this thing is used. The composite logic remains intact. 
</blockquote></details>

Oh, I see, the comment is bad. I'll update it. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo, @mgartner, and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1115 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7Ilp_57YvzQAtSDcTx:-N7IoLfS9sRXGnYGS_6Z:bswwtq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/864b7d9c28c68286403926156965a641cfb90901/pkg/sql/rowenc/index_encoding.go#L1115)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Oh, I see, the comment is bad. I'll update it. 
</blockquote></details>

Alright, I've reworded the comment. Hopefully it's clearer now. 


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
Fourth and fifth commit LGTM.

> Reviewed 1 of 1 files at r9, 2 of 2 files at r10.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7IslXwEIwLDnsIF5LA:b64hz2c)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @livlobo and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1115 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7Ilp_57YvzQAtSDcTx:-N7IsESk7wVx7PMF9m51:b-n9d8fp) ([raw file](https://github.com/cockroachdb/cockroach/blob/864b7d9c28c68286403926156965a641cfb90901/pkg/sql/rowenc/index_encoding.go#L1115)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Alright, I've reworded the comment. Hopefully it's clearer now. 
</blockquote></details>

Brilliant, thank you.


<!-- Sent from Reviewable.io -->

fqazi reviewed (changes requested):
For end-to-end testing are we comfortable adding multistatement transactions? i.e. Two drop columns, I know we can't do a mix of ADD/DROP yet?

Looks good, but didn't give enough time to the builder and some test stuff. I'll do that tomorrow

> Reviewed 2 of 74 files at r4, 82 of 82 files at r5, 28 of 28 files at r6, 5 of 5 files at r7, 1 of 1 files at r9, 2 of 2 files at r10, 35 of 35 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7IuBsN9Ak0IZoayn_7:b-5373qt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @livlobo, and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1131 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7ItjHf6J8TMbS2YRZP:-N7ItjHf6J8TMbS2YRZQ:b-eb0hmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0fbc3d81c681bd35b97feb291043dbff596e30b/pkg/sql/rowenc/index_encoding.go#L1131)):*
> ```Go
> 	// columns set is not populated and instead we defer to the colMap to compute
> 	// the complete set before subtracting the key columns.
> 	if index.GetVersion() < descpb.PrimaryIndexWithStoredColumnsVersion {
> ```

I guess we will infinitely support older encoding types? Until indexes get recreated.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 164 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7IvhEJ8YjksZngb8dC:-N7IvhEJ8YjksZngb8dD:b-kyt03p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5785ab8a36ab30dbbd7422863cdc8ffca81bfc8e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L164)):*
> ```Go
> 		b.Drop(cc)
> 	}
> 	handleDropColumnDefaultExpression(b, col, colElts, behavior)
> ```

This pattern is much cleaner than what we have in add column, we should do the same there in the future.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 307 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7IwoQzCCpWSOXIQZks:-N7IwoQzCCpWSOXIQZkt:bducpn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/5785ab8a36ab30dbbd7422863cdc8ffca81bfc8e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L307)):*
> ```Go
> }
> 
> func toPublicNotCurrentlyPublicFilter(
> ```

Would it make sense to just have a routine that returns a filter closure?

statusAndTargetFilter( scpb.Status_PUBLIC,  scpb.ToPublic)
statusFilter(...)

Just a thought more cosmetic I think.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/column.go` line 123 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7IsVwi1aRQMARPov5l:-N7IsVwj33n_IVyZFnxk:bb0tell) ([raw file](https://github.com/cockroachdb/cockroach/blob/651e50bbb505f96259edd543296b6bd471c430e6/pkg/sql/schemachanger/scexec/scmutationexec/column.go#L123)):*
> ```Go
> 		*(protoutil.Clone(mut.GetColumn())).(*descpb.ColumnDescriptor))
> 
> 	// Ensure that the column is added in the right location. This is important
> ```

Is there a reason we couldn't push this ordering into the dep rules? I guess removing or adding stuff in the middle?
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 503 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7It2qN6OO5Uk0chujo:-N7It2qN6OO5Uk0chujp:boowv8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/651e50bbb505f96259edd543296b6bd471c430e6/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L503)):*
> ```Go
> 	// a column which does not exist. This would lead to panics inside the
> 	// optimizer and an invalid table descriptor.
> 	registerDepRule("indexes containing columns reach absent before column",
> ```

I guess validation would also catch this? Or it might be something should check for?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> For end-to-end testing are we comfortable adding multistatement transactions? i.e. Two drop columns, I know we can't do a mix of ADD/DROP yet?

I added a few of these types of tests. They were good. Caught a bug with improperly sorting the statements. Also figured out I had not handled dropping the OnUpdate expression. It was good to flex the new data-driven explains as well. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7Lj8m25lpmjsguVySQ:b-1nuyvp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @livlobo, and @postamar)

___
*[`pkg/sql/rowenc/index_encoding.go` line 1131 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7ItjHf6J8TMbS2YRZP:-N7LiWbB4dNjxwpQaQgi:btecd64) ([raw file](https://github.com/cockroachdb/cockroach/blob/a0fbc3d81c681bd35b97feb291043dbff596e30b/pkg/sql/rowenc/index_encoding.go#L1131)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I guess we will infinitely support older encoding types? Until indexes get recreated.
</blockquote></details>

Yep. At least we know they aren't created by the declarative schema changer.
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/column.go` line 123 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7IsVwi1aRQMARPov5l:-N7LihzT-SiowkD3603l:b-gdukn5) ([raw file](https://github.com/cockroachdb/cockroach/blob/651e50bbb505f96259edd543296b6bd471c430e6/pkg/sql/schemachanger/scexec/scmutationexec/column.go#L123)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Is there a reason we couldn't push this ordering into the dep rules? I guess removing or adding stuff in the middle?
</blockquote></details>

We won't necessarily have elements in the plan for the other columns which aren't being dropped. Consider the case of rolling back the drop column of the first column in the table. 
___
*[`pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go` line 503 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7It2qN6OO5Uk0chujo:-N7Lix6dAwL2GZ6TmUR5:b-1woxv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/651e50bbb505f96259edd543296b6bd471c430e6/pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go#L503)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I guess validation would also catch this? Or it might be something should check for?
</blockquote></details>

I'll add the validation in a separate PR. 


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
This is RFAL

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7Mp1WM82Vh_LY9m8Cm:b-n27ya0)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @livlobo, @mgartner, and @postamar)

___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go` line 307 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7IwoQzCCpWSOXIQZks:-N7MachtCIR3nX6gru0x:b-r4x3t2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5785ab8a36ab30dbbd7422863cdc8ffca81bfc8e/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/helpers.go#L307)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Would it make sense to just have a routine that returns a filter closure?

statusAndTargetFilter( scpb.Status_PUBLIC,  scpb.ToPublic)
statusFilter(...)

Just a thought more cosmetic I think.
</blockquote></details>

I didn't do anything in this commit but I'm interested in improvements here in future commits. 


<!-- Sent from Reviewable.io -->

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_primary_index.go:
```diff
@@ -62,6 +61,20 @@ func init() {
 					}
 				}),
 			),
+			// The transition from MERGED to WRITE_ONLY must precede index validation.
+			// In MERGE_ONLY and MERGED, the index receives writes, but writes do not
+			// enforce uniqueness (they don't use CPut, see ForcePut). In WRITE_ONLY,
+			// the index receives normal writes. Only once writes are enforcing the
+			// uniqueness constraint for new can we validate that the index does
+			// indeed represent a valid uniqueness constraint over all rows.
+			to(scpb.Status_WRITE_ONLY,
+				emit(func(this *scpb.PrimaryIndex) *scop.MakeMergedIndexWriteOnly {
+					return &scop.MakeMergedIndexWriteOnly{
+						TableID: this.TableID,
+						IndexID: this.IndexID,
+					}
+				}),
+			),
```
This is a big change! I previously don't know that an index that is in MERGE_ONLY or MERGED can still receive writes. What is the concern of checking uniqness of writes during merging previously (when we transition the primary index to WRITE_ONLY before merging with the temporary index)?

ajwerner replied:
While the index is in `MERGING` in the table descriptor, write to it use `Put` instead of `CPut`. See `ForcePut` on `catalog.Index`. 

Consider the following timeline:

create table t (i int primary key, j int);
insert into t values (1, 1);
create unique index idx on t(j);

t1: index merging concludes
at this point the unique index contains:
`1: 1`

t2: validation begins by reading a snapshot at t2 to count that the number of rows in the unique index is equal to the number of rows in the primary index
t3: `insert into t values (2, 1)` -- this will lead to the unique index now only containing `1: 2` instead of `1: 1` and the force from `MERGING` means the write will succeed.
t4: Validation completes successfully and the index is moved to public

At this point we have a corrupt unique secondary index.

If, however, we moved from `MERGING` to `WRITE_ONLY` before proceeding with validation, the insert operation would fail and the index will remain consistent.

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -497,6 +497,32 @@ func init() {
+				joinOnColumnID(to.el, ct, "table-id", "column-id"),
+				targetStatusEq(from.target, to.target, scpb.ToAbsent),
+				currentStatusEq(from.node, to.node, scpb.Status_ABSENT),
+				rel.Filter("columnTypeIsNotBeingDropped", ct)(func(
+					ct *scpb.ColumnType,
+				) bool {
+					return !ct.IsRelationBeingDropped
+				}),
```
1. I don't quite understand the purpose of `ct` here, which is a ColumnType element?
2. Why do we need this clause that uses the `isRelationBeingDropped` field, which I thought was deemed "hacky"?

postamar replied:
I think this is needed to gracefully handle `DROP TABLE`. IIRC we wanted to replace this filter (and thus remove this field) by using better rule constructs, something along the lines of "this rule only applies on this column if the parent table isn't present in the rel-database with an ABSENT target".

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_index_and_column.go:
```diff
@@ -784,6 +812,29 @@ func init() {
 				}),
 			}
 		})
+
+	// We need to ensure that the temporary index has all the relevant writes
+	// to any columns it contains. We ensure elsewhere that any index which
+	// will later be merged with the temporary index is not backfilled until
+	// that temporary index is receiving writes. This rule ensures that those
+	// write operations contain data for all columns.
+	registerDepRule(
+		"column is WRITE_ONLY before temporary index is WRITE_ONLY",
+		scgraph.Precedence,
+		"column", "index",
+		func(from, to nodeVars) rel.Clauses {
+			return rel.Clauses{
+				from.el.Type((*scpb.Column)(nil)),
+				to.el.Type((*scpb.TemporaryIndex)(nil)),
+				indexContainsColumn(to.el, from.el, "index-column", "table-id", "column-id", "index-id"),
+				targetStatus(from.target, scpb.ToPublic),
+				targetStatus(to.target, scpb.Transient),
+				currentStatus(from.node, scpb.Status_WRITE_ONLY),
+				currentStatus(to.node, scpb.Status_WRITE_ONLY),
+			}
+		},
+	)
+
```
This rule seems to be relevant when adding a column (so we insist this newly added column reaches WRITE_ONLY before the temporary reaches WRITE_ONLY). I'm curious how you discover the need to add such a dep rule when working on `DROP COLUMN`

ajwerner replied:
You're right that this is unrelated to drop column. I don't remember where this came up, but there was a case I ran into where this mattered. I think it was that I removed the `WRITE_ONLY` state from `"column depends on primary index"` because it broke rollbacks of drop column, at which point in the add column case now it was valid to defer moving the column to `WRITE_ONLY` until later which caused some other problem.

ajwerner commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_secondary_index.go:
```diff
@@ -46,14 +46,13 @@ func init() {
 				}),
 			),
 			to(scpb.Status_MERGE_ONLY,
```
We always need to merge concurrent writes into an index when backfilling as of 22.1. The reasons for this are somewhat surprising and relate to history rewriting. See this rfc: https://github.com/cockroachdb/cockroach/blob/master/docs/RFCS/20211004_incremental_index_backfiller.md

fqazi reviewed (changes requested):
Looks good just a few minor questions from me. I think most of those are non-issues.

> Reviewed 71 of 71 files at r12, 28 of 28 files at r13, 5 of 5 files at r14, 1 of 1 files at r15, 2 of 2 files at r16, 49 of 49 files at r17, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7RShbR6fvfWbFjJfti:b-dbg111)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @livlobo, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/BUILD.bazel` line 12 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7R7zsqAcNR6RRWokND:-N7R7zsqAcNR6RRWokNE:bhybgit) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/BUILD.bazel#L12)):*
> ```Starlark
> go_test(
>     name = "schemachanger_test",
>     size = "large",
> ```

I guess the drop stuff ballooned the length of this test, hopefully we can get this down with parallelism.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 282 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RRoNn84ksclyZRG8D:-N7RRoNn84ksclyZRG8E:b-8khut8) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L282)):*
> ```Go
> 		}
> 		if ic.Kind != scpb.IndexColumn_STORED {
> 			panic(errors.AssertionFailedf("can only drop columns which are stored in the primary index, this one is %v ",
> ```

Are we surfacing the right error here? Or should this be "column %q is referenced by the primary key". Just sanity checking
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 353 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RPzGp6-nG0EQj84Bs:-N7RPzGp6-nG0EQj84Bt:b-wf7g1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L353)):*
> ```Go
> 	for i := n + 1; i < len(storedColumns); i++ {
> 		storedColumns[i].OrdinalInKind--
> 		b.Add(storedColumns[i])
> ```

I think this is fine because it will just replace the existing element based on the remaining keys. I'm guessing the ops will do the heavy lifting here since the target state will stay the same
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 374 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RQRmKDlz5dWuu5v1t:-N7RQRmKDlz5dWuu5v1u:b-9846e7) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L374)):*
> ```Go
> 		return
> 	}
> 	// TODO(ajwerner): Support dropping UNIQUE WITHOUT INDEX constraints.
> ```

We might just want a checklist-type Github issue for these. As we are doing this type of stuff over time some of these cases will take longer to address.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 458 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RKOykAyjO7iYS5JCQ:-N7RKOykAyjO7iYS5JCR:b-3v6o0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L458)):*
> ```Go
> 
> func handleDropColumnExpressions(b BuildCtx, colElts ElementResultSet, behavior tree.DropBehavior) {
> 	publicTargets := colElts.Filter(publicTargetFilter)
> ```

Does it make sense for the caller to give a filtered set of elements?
___
*[`pkg/sql/schemachanger/scexec/scmutationexec/column.go` line 123 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7IsVwi1aRQMARPov5l:-N7R6EUR0CVugvRFsqIT:bmrsidb) ([raw file](https://github.com/cockroachdb/cockroach/blob/651e50bbb505f96259edd543296b6bd471c430e6/pkg/sql/schemachanger/scexec/scmutationexec/column.go#L123)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

We won't necessarily have elements in the plan for the other columns which aren't being dropped. Consider the case of rolling back the drop column of the first column in the table. 
</blockquote></details>

Ooh right


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
@fqazi other than the passing of the public elements are there any changes requested? Can I defer that? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7RYbtR5mtdzX1iZ2R8:b-te6p34)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @livlobo, @postamar, and @Xiang-Gu)

___
*[`pkg/sql/schemachanger/BUILD.bazel` line 12 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7R7zsqAcNR6RRWokND:-N7RYUHMA4pAJD2rTORW:b91s5d3) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/BUILD.bazel#L12)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I guess the drop stuff ballooned the length of this test, hopefully we can get this down with parallelism.
</blockquote></details>

Indeed. I'm waiting on Ricky's logictest change before doing something. 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 282 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RRoNn84ksclyZRG8D:-N7RY6lg-BGLo-agkW8r:b-xv3gsp) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L282)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Are we surfacing the right error here? Or should this be "column %q is referenced by the primary key". Just sanity checking
</blockquote></details>

This is an assertion failure because we already checked whether it was a primary key column earlier in the `dropColumn` function in `checkColumnNotInPrimaryKey`. I don't expect to see this error.
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 353 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RPzGp6-nG0EQj84Bs:-N7RY1dL89OxpBVIqhHy:b-ttscfb) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L353)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

I think this is fine because it will just replace the existing element based on the remaining keys. I'm guessing the ops will do the heavy lifting here since the target state will stay the same
</blockquote></details>

Yes, we'll overwrite the existing element being added. 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 374 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RQRmKDlz5dWuu5v1t:-N7RYR-S9NF9SBlyC1lO:b-qwwo00) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L374)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

We might just want a checklist-type Github issue for these. As we are doing this type of stuff over time some of these cases will take longer to address.
</blockquote></details>

Yes, I'll create one. 
___
*[`pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go` line 458 at r17](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-N7RKOykAyjO7iYS5JCQ:-N7RYYIX9N397JvCy1vc:b2e4k81) ([raw file](https://github.com/cockroachdb/cockroach/blob/07ce7b155892127bde9fff3d976897419d6cd97a/pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go#L458)):*
<details><summary><i>Previously, fqazi (Faizan Qazi) wrote…</i></summary><blockquote>

Does it make sense for the caller to give a filtered set of elements?
</blockquote></details>

I suppose, but I'm not eager to wait for another CI cycle if you're amenable. 


<!-- Sent from Reviewable.io -->

fqazi reviewed (approved):
Yes you can defer that <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84563#-:-N7RZ-9V8D14kbitUJbM:b-czhex4)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @ajwerner, @fqazi, @livlobo, @postamar, and @Xiang-Gu)



<!-- Sent from Reviewable.io -->

ajwerner commented:
bors r+
